File: python/attp/core/agent_tracer.py
"""Agent-side tracer facade — composition of KeyStore + ChainManager.

Agent components (ATTPClient, ATTPServer, MCPToolBridge) only need
hop construction and private-key access.  No database dependency.
"""

from __future__ import annotations

from attp.core.authentication import KeyStore
from attp.core.provenance import ChainManager


class AgentTracer:
    """Lightweight facade for Agent-side tracing operations.

    Combines KeyStore (key caching) and ChainManager (hop construction)
    without any persistent storage.
    """

    def __init__(self) -> None:
        self._key_store = KeyStore()
        self._chain = ChainManager(key_store=self._key_store)

    # -- key management (delegated to KeyStore) --

    def cache_public_key(self, node_did: str, public_key) -> None:
        """Inject a public key into cache (called before verification)."""
        self._key_store.cache_public_key(node_did, public_key)

    def load_private_key(self, key_path: str):
        """Load a private key by file path (cached)."""
        return self._key_store.load_private_key(key_path)

    # -- chain operations (delegated to ChainManager) --

    def append_hop(self, metadata: dict, content: str, node_did: str,
                   target_did: str, private_key_path: str) -> dict:
        return self._chain.append_hop(
            metadata, content, node_did, target_did, private_key_path,
        )

    # -- internal access (for DIDResolver injection) --

    @property
    def key_store(self) -> KeyStore:
        """Public read-only access to the internal KeyStore."""
        return self._key_store

File: typescript/attp/core/agent-tracer.ts
/**
 * Agent-side tracer facade — 组合 KeyStore + ChainManager。
 *
 * 从 python/attp/core/agent_tracer.py 忠实移植。
 *
 * Agent 侧仅使用 appendHop + cachePublicKey：
 *   - Python 的 load_private_key 在 TS 端不需要——openclaw 通道在启动时
 *     调用 loadPrivateKeyPem 一次性加载私钥，并将 LoadedKey 直接传给 appendHop。
 *   - 因此此处不暴露 load_private_key（避免误导；openclaw 直接持有 LoadedKey）。
 */

import { KeyStore, type LoadedKey, type AnyKey } from "./authentication/keys.js";
import { ChainManager, type ChainHopMetadata } from "./provenance/chain.js";

/**
 * 轻量级 facade：Agent 侧溯源操作。
 *
 * 组合 KeyStore（公钥缓存）与 ChainManager（跳构造），不持有任何持久化存储。
 * 镜像 Python AgentTracer：构造时创建独立 KeyStore 并将其注入 ChainManager。
 */
export class AgentTracer {
  readonly keys: KeyStore;
  readonly chain: ChainManager;

  constructor() {
    // 使用构造体而非字段初始化器：先建 keys 再注入 chain，语义与 Python __init__ 一致，
    // 同时规避在字段初始化器中引用 this.keys 的初始化顺序歧义。
    this.keys = new KeyStore();
    this.chain = new ChainManager(this.keys);
  }

  /**
   * 注入公钥到缓存（验签前调用）。
   * 委托给内部 KeyStore。
   */
  cachePublicKey(did: string, key: LoadedKey): void {
    this.keys.cachePublicKey(did, key);
  }

  /**
   * 追加一跳：计算哈希、签名、写入 recordedHop，返回新的 metadata。
   *
   * 调用方传入已加载的私钥（openclaw 启动时加载一次）。
   * 委托给内部 ChainManager。
   */
  appendHop(
    metadata: ChainHopMetadata,
    content: string,
    senderDid: string,
    targetDid: string,
    privateKey: AnyKey,
    behaviorType?: string,
  ): Promise<ChainHopMetadata> {
    return this.chain.appendHop(
      metadata,
      content,
      senderDid,
      targetDid,
      privateKey,
      behaviorType,
    );
  }
}
